dpnp/backend/extensions/lapack/getrf.cpp

```diff
@@ -0,0 +1,200 @@
+                              char *in_a,
+                              std::int64_t lda,
+                              std::int64_t *ipiv,
+                              std::int64_t *dev_info,
+                              std::vector<sycl::event> &host_task_events,
+                              const std::vector<sycl::event> &depends)
+{
+    type_utils::validate_type_for_device<T>(exec_q);
+
+    T *a = reinterpret_cast<T *>(in_a);
+
+    const std::int64_t scratchpad_size =
+        mkl_lapack::getrf_scratchpad_size<T>(exec_q, n, n, lda);
+    T *scratchpad = nullptr;
+
+    std::stringstream error_msg;
+    std::int64_t info = 0;
+
+    sycl::event getrf_event;
+    try {
+        scratchpad = sycl::malloc_device<T>(scratchpad_size, exec_q);
+
+        getrf_event = mkl_lapack::getrf(
+            exec_q,
+            n,    // The order of the square matrix A (0 ≤ n).
+                  // It must be a non-negative integer.
+            n,    // The number of columns in the square matrix A (0 ≤ n).
+                  // It must be a non-negative integer.
+            a,    // Pointer to the square matrix A (n x n).
+            lda,  // The leading dimension of matrix A.
+                  // It must be at least max(1, n).
+            ipiv, // Pointer to the output array of pivot indices.
+            scratchpad, // Pointer to scratchpad memory to be used by MKL
+                        // routine for storing intermediate results.
+            scratchpad_size, depends);
+    } catch (mkl_lapack::exception const &e) {
+        error_msg
+            << "Unexpected MKL exception caught during getrf() call:\nreason: "
+            << e.what() << "\ninfo: " << e.info();
+        info = e.info();
+    } catch (sycl::exception const &e) {
+        error_msg << "Unexpected SYCL exception caught during getrf() call:\n"
+                  << e.what();
+        info = -1;
+    }
+
+    if (info != 0) // an unexpected error occurs
+    {
+        if (scratchpad != nullptr) {
+            sycl::free(scratchpad, exec_q);
+        }
+
+        if (info < 0) {
+            throw std::runtime_error(error_msg.str());
+        }
+    }
+
+    sycl::event write_info_event = exec_q.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(getrf_event);
+        cgh.single_task([=]() { dev_info[0] = info; });
+    });
+
+    host_task_events.push_back(write_info_event);
+
+    sycl::event clean_up_event = exec_q.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(getrf_event);
+        auto ctx = exec_q.get_context();
+        cgh.host_task([ctx, scratchpad]() { sycl::free(scratchpad, ctx); });
+    });
+    host_task_events.push_back(clean_up_event);
+    return getrf_event;
+}
+
+std::pair<sycl::event, sycl::event>
+    getrf(sycl::queue q,
+          dpctl::tensor::usm_ndarray a_array,
+          dpctl::tensor::usm_ndarray ipiv_array,
+          dpctl::tensor::usm_ndarray dev_info_array,
+          const std::int64_t n,
```

Comment:
Yes, you are right but I pass it in to avoid re-calculating it.
If I calculate it in python function, why should I calculate it in backend if I can pass i?

Reply:
done